consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContext.java

```diff
@@ -0,0 +1,290 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Iterables;
+
+/*
+Responsible for creating an environment in which integration testing can be conducted.
+
+The test setup is a 'n' node network, the first of which can be considered the "local" node.
+All allValidators are in the genesis block's extraData. Validator
+
+ */
+public class TestContext {
+
+  public static final int EPOCH_LENGTH = 10_000;
+  public static final int BLOCK_TIMER_SEC = 3;
+  public static final int ROUND_TIMER_SEC = 12;
+
+  private Map<Address, ValidatorPeer> remotePeers;
+  private final MutableBlockchain blockchain;
+  private final IbftController controller;
+  private final IbftFinalState finalState;
+
+  public TestContext(
+      final Map<Address, ValidatorPeer> remotePeers,
+      final MutableBlockchain blockchain,
+      final IbftController controller,
+      final IbftFinalState finalState) {
+    this.remotePeers = remotePeers;
+    this.blockchain = blockchain;
+    this.controller = controller;
+    this.finalState = finalState;
+  }
+
+  public Collection<ValidatorPeer> getRemotePeers() {
+    return remotePeers.values();
+  }
+
+  public MutableBlockchain getBlockchain() {
+    return blockchain;
+  }
+
+  public IbftController getController() {
+    return controller;
+  }
+
+  public MessageFactory getLocalNodeMessageFactory() {
+    return finalState.getMessageFactory();
+  }
+
+  public Block createBlockForProposal(final int round, final long timestamp) {
+    return finalState
+        .getBlockCreatorFactory()
+        .create(blockchain.getChainHeadHeader(), round)
+        .createBlock(timestamp);
+  }
+
+  public RoundSpecificNodeRoles getRoundSpecificRoles(final ConsensusRoundIdentifier roundId) {
+    // This will return NULL if the LOCAL node is the proposer for the specified round
+    final Address proposerAddress = finalState.getProposerForRound(roundId);
+    final ValidatorPeer proposer = remotePeers.getOrDefault(proposerAddress, null);
+
+    final List<ValidatorPeer> nonProposers = new ArrayList<>(remotePeers.values());
+    nonProposers.remove(proposer);
+
+    return new RoundSpecificNodeRoles(proposer, remotePeers.values(), nonProposers);
+  }
+
+  public NodeParams getLocalNodeParams() {
+    return new NodeParams(finalState.getLocalAddress(), finalState.getNodeKeys());
+  }
+
+  public static TestContext createTestEnvWithUtcClock(
+      final int validatorCount, final int indexOfFirstLocallyProposedBlock) {
+    return createTestEnvironment(
+        validatorCount, indexOfFirstLocallyProposedBlock, Clock.systemUTC());
+  }
+
+  public static TestContext createTestEnvironment(
+      final int validatorCount, final int indexOfFirstLocallyProposedBlock, final Clock clock) {
+
+    final NetworkLayout networkNodes =
```

Comment:
this method is pretty nasty, haven't tried yet but is there any reasonable way to break it up a bit?

Reply:
Heh - this is after my first attempt to break it up. See what I can do.